docs/virtual-package.md

```diff
@@ -0,0 +1,9 @@
+# Composer Virtual Packages
+
+Virtual packages are a way to specify the dependency on an implementation of an interface-only repository without forcing a specific implementation. For Httplug, the virtual package is called `php-http/client-implementation`. 
+
+There is no project registered with that name. However, all client implementations including client adapters for Httplug use the `provide` section to tell composer that they do provide the client-implementation.
+
+When *writing a reusable library*, the `require` section should only mention `php-http/client-implementation`, while the `require-dev` section needs to specify a concrete implementation in most cases, in order for the package to be installable on its own for development and testing.
+
+When *using a reusable library* in an application, the `require` section needs to include a concrete implementation in order for the application to install. Missing to do that will lead to composer reporting a missing package.
```

Comment:
I agree, it is a bit weird. I would close to the problem from a library point of view: libraries does not depend on an actual implementation, but a virtual one. This is to avoid hard coupling and to allow using an implementation chosen by the user. Applications however must require one providing that virtual package. You can think about this as an "interface/contract" for packages.

I don't think including the composer error message adds any value. This is a composer detail. The actual message doesn't even matter as it leads directly to the fix, so it does not help identifying the problem, it's itself the identification. Enough knowledge to use composer is required.